File: online_exam/exam/static/js/mcqQuestionView.js
$(document).ready(function(){
    $(document).ajaxStart($.blockUI).ajaxStop($.unblockUI);
    //$(document).ajaxStart($.blockUI({ message: $('#throbber') })).ajaxStop($.unblockUI);
    $('#mcqQuesEditTbl').hide();
    $('#mcqSubmitBtn').prop('disabled',true);

    $('#mcqLink').on('click',function(){
        $.ajax({
            url: "/subject_list/",
            type: "GET",
            dataType:"JSON",
            success: function(response) {
                //console.log(response);
                $("#mcqSubjectList").find('option').not(':first').remove();
                if(response.length!=0){
                    $.each(response.results,function(index,row){
                        //$.each(row, function(k, v){
                            //console.log(v.subject_id);
                            $("#mcqSubjectList")
                            .append(
                                $("<option></option>")
                                    .text(row.subject_id+':'+row.subject_name)
                                    .val(row.subject_id)
                            );
                        //});
                    });
                }
            }
        });
    });
    $('#mcqSubjectList').on('change',function(){
        $("#mcqQuesTbl").find('tr').not(':first').remove();
        $("#mcqTestList").prop('selectedIndex',0);
        //console.log("Test");
        $.ajax({
            url: "/test_list/",
            type: "GET",
            data: {'subject_id':$('#mcqSubjectList').val(),
                   'test_type':1},
            dataType:"JSON",
            success: function(response) {
                var data=response;
                $("#mcqTestList").find('option').not(':first').remove();
                if(response.length!=0){
                    $.each(response,function(index,row){
                        $.each(row, function(k, v){
                            $("#mcqTestList")
                            .append(
                                $("<option></option>")
                                    .text(v.test_id.toString().substring(3, 5)+': '+v.test_name)
                                    .val(v.test_id)
                            );
                        });
                    });
                }
            }
        });
    });

    $('#mcqTestList').on('change',function(){
        $("#mcqQuesTbl").find('tr').not(':first').remove();
        $.ajax({
            url: "/test_info/",
            type: "GET",
            data: {'subject_id':$('#mcqSubjectList').val(),
                   'test_type':1,
                   'test_id' : $('#mcqTestList').val()},
            dataType:"JSON",
            success: function(response) {
                $('#mcqMaxQuesNo').text(response.results[0].test_total_questions);
                $('#mcqExamTime').text(response.results[0].test_totaltimes+" mins");
                $('#mcqTotalMarks').text(response.results[0].test_totalmarks);
                //mcq_available_check();
                get_mcq_ques_data();
            }
        });
    });

    $('#mcqQuesSubBtn').on('click',function(){
        $('#mcqSubjectList').prop('disabled',true);
        $('#mcqTestList').prop('disabled',true);
        $('#mcqQuesTbl').show();
        $('#mcqQuesEditTbl').hide();
        $('#mcqSubmitBtn').prop('disabled',false);
        $("#mcqQuesTbl").find('tr').not(':first').remove();
        var quesLimit=$("#mcqMaxQuesNo").text();
        var trHTML = '';
        for(var i=1;i<=quesLimit;i++){
            trHTML+= '<tr id='+i+'><td><span>'+i+'</span><input type="text" class="form-control hidden" id="qid" name="qid" value="'+i+'"></td>'+
//            '<td><input type="text" class="form-control" id="ques" name="ques" placeholder="Question"></td>'+
            '<td><textarea class="form-control vresize quesTest" id="ques" name="ques" placeholder="Question"></textarea></td>'+
            '<td><textarea class="form-control vresize" id="option1" name="option1" placeholder="Option 1"></textarea></td>'+
            '<td><textarea class="form-control vresize" id="option2" name="option2" placeholder="Option 2"></textarea></td>'+
            '<td><textarea class="form-control vresize" id="option3" name="option3" placeholder="Option 3"></textarea></td>'+
            '<td><textarea class="form-control vresize" id="option4" name="option4" placeholder="Option 4"></textarea></td>'+
            /*'<td><input type="text" class="form-control" id="option1" name="option1" placeholder="Option 1"></td>'+
            '<td><input type="text" class="form-control" id="option2" name="option2" placeholder="Option 2"></td>'+
            '<td><input type="text" class="form-control" id="option3" name="option3" placeholder="Option 3"></td>'+
            '<td><input type="text" class="form-control" id="option4" name="option4" placeholder="Option 4"></td>'+*/
            //'<td><input type="text" class="form-control" id="rightAnswer" name="rightAnswer" placeholder="Right Answer" required></td></tr>';
            '<td><select class="form-control rightAnswer" name="rightAnswer"></select></td></tr>';
        }
        $('#mcqQuesTbl').append(trHTML);
        getRightAnsOptions();
    });


    var mcqForm = $( "#mcqSectionForm" );
    mcqForm.validate();

    $('#mcqSubmitBtn').on('click',function(){
        if(mcqForm.valid()) {
            var data=[];
            $('#mcqQuesTbl').find('tr:not(:has(th))').each(function(){
                var id=$(this).attr('id');
                var row={};
                $(this).find('textarea, :input').each(function(){
                    row[$(this).attr('name')]=$(this).val();
                });
                data.push(row);
            });
            console.log(data);
            if (data.length === 0) {
                swal("Error!", "Please generate the form!", "error");
            }
            else{
                $.ajax({
                    url: "/mcq_paper_submit/",
                    type: "POST",
                    data: {obj:JSON.stringify(data),
                    'csrfmiddlewaretoken' : $("input[name=csrfmiddlewaretoken]").val(),
                    'test_id':$('#mcqTestList').val()},
                    dataType:"JSON",
                    success: function(response) {
                        //console.log(response);

                        console.log(response.status);
                        if(response.status=='1'){
                            swal("Success!", "Successfully Submitted!", "success");
                            resetMCQQuesSection();
                        }
                        else{
                            swal("Error!", "Something Wrong!", "error");
                            resetMCQQuesSection();
                        }
                    }
                });
            }
        }
    });

    function mcq_available_check(){
        $('#mcqSubmitBtn').prop('disabled',true);
        $.ajax({
            url: "/mcq_available_check/",
            type: "GET",
            data: {'test_id':$('#mcqTestList').val()},
            dataType:"JSON",
            success: function(response) {
                if(response){
                    $("#mcqQuesEditTbl").find('tr').not(':first').remove();
                    var trHTML = '';
                    $.each(response, function (key,value) {
                        trHTML +='<tr data-id="'+value.mcq_question_id+'" data-ques="'+value.mcq_question+'" data-option1="'+value.mcq_option1+'" data-option2="'+value.mcq_option2+'" data-option3="'+value.mcq_option3+'" data-option4="'+value.mcq_option4+'" data-ans="'+value.mcq_right_answer+'">'+
                        '<td data-id="'+value.mcq_question_id+'">' + value.mcq_question_id +'</td>'+
                        '<td>' + value.mcq_question.replace(/\n/g, "<br />") +'</td>'+
                        '<td>' + value.mcq_option1 +'</td>'+
                        '<td>' + value.mcq_option2 +'</td>'+
                        '<td>' + value.mcq_option3 +'</td>'+
                        '<td>' + value.mcq_option4 +'</td>'+
                        '<td>' + value.mcq_right_answer +'</td></tr>';
                    });
                    $('#mcqQuesEditTbl').append(trHTML);
                }
            }
        });
    }

    $('#mcqQuesEditTbl').on('click','tr', function(){
        $('#quesEditModal').modal('show');
        $('#quesIDEdit').val($(this).attr("data-id"));
        $('#mcqQuesEdit').val($(this).attr("data-ques"));
        $('#mcqOptionOneEdit').val($(this).attr("data-option1"));
        $('#mcqOptionTwoEdit').val($(this).attr("data-option2"));
        $('#mcqOptionThreeEdit').val($(this).attr("data-option3"));
        $('#mcqOptionFourEdit').val($(this).attr("data-option4"));
        $('#mcqRightAnswerEdit').val($(this).attr("data-ans"));
    });

    $('#mcqQuesEditBtn').on('click',function(){
        $.unblockUI();
        $('#quesEditModal').modal('hide');
        $.ajax({
            url: "/mcq_ques_edit/",
            type: "GET",
            data: {'test_id': $('#mcqTestList').val(),
                    'mcq_ques_id': $('#quesIDEdit').val(),
                    'mcq_ques': $('#mcqQuesEdit').val(),
                    'mcq_option1': $('#mcqOptionOneEdit').val(),
                    'mcq_option2': $('#mcqOptionTwoEdit').val(),
                    'mcq_option3': $('#mcqOptionThreeEdit').val(),
                    'mcq_option4': $('#mcqOptionFourEdit').val(),
                    'mcq_answer': $('#mcqRightAnswerEdit').val()
                    },
            dataType:"JSON",
            success: function(response) {
                if(response.status==1){
                    swal("Success!", "Successfully Updated!", "success");
                    mcq_available_check();
                }
                else{
                    swal("Error!", "Something Wrong!", "error");
                }
            }
        });
    });

    function get_mcq_ques_data(){
        $.ajax({
            url: "/get_mcq_ques_data/",
            type: "GET",
            data: {'test_id': $('#mcqTestList').val()},
            dataType:"JSON",
            success: function(response) {
                //console.log(response.status);
                if(response.status==0){
                    $('#mcqEditBtn').prop('disabled',true);
                    $('#mcqQuesSubBtn').prop('disabled',false);
                }
                else{
                    $('#mcqEditBtn').prop('disabled',false);
                    $('#mcqQuesSubBtn').prop('disabled',true);
                }
            }
        });
    }

    $('#mcqEditBtn').on('click',function(){
        $('#mcqQuesEditTbl').show();
        $('#mcqQuesTbl').hide();
        $('#mcqSubjectList').prop('disabled',true);
        $('#mcqTestList').prop('disabled',true);
        mcq_available_check();
    });

    function getRightAnsOptions(){
        $('.rightAnswer').each(function(index,row){
            for(var i=1; i<= 4; i++){
                 $(this).append($("<option></option>").text(i).val(i));
            }
        });
    }

    /*$('#resetBtn').on('click',function(){
        resetTestSection();
        enable();
        $('#quesNo').text("");
        $("#totalMarks").prop('readonly',true);
        $("#submitTestBtn").prop('disabled',false);
        $("#editTestBtn").prop('disabled',true);
    });*/

    function resetMCQQuesSection(){
        var validator = $( "#mcqSectionForm" ).validate();
        validator.resetForm();
        $("input[type=text]").val("");
        $("textarea").val("");
        $("select").prop('selectedIndex',0);
        $(":input").closest('.form-group').removeClass('has-success');
        $("#mcqQuesTbl tbody>tr").remove();
        $("#mcqQuesEditTbl tbody>tr").remove();
        $('#mcqMaxQuesNo').text('');
        $('#mcqExamTime').text('');
        $('#mcqTotalMarks').text('');
        $('#mcqSubjectList').prop('disabled',false);
        $('#mcqTestList').prop('disabled',false);
        $('#mcqSubmitBtn').prop('disabled',true);
    }

    $('#mcqResetBtn').on('click',function(){
        resetMCQQuesSection();
        $('#mcqEditBtn').prop('disabled',true);
        $('#mcqQuesSubBtn').prop('disabled',true);
    });

});

File: online_exam/exam/views/McqQuestionView.py
import json

from django.shortcuts import render,get_object_or_404
from django.http import HttpResponse, JsonResponse
import datetime

from exam.models import Subject
from exam.models import Test
from exam.models import McqQuestion


def index(request):
    return render(request,'exam/onlineExam.html',{})

def getAllTestList(request):
    if (request.method == "GET"):
        test_type = request.GET.get('test_type')
        subject_id = request.GET.get('subject_id')
        testList = Test.objects.values('test_id','test_name','test_totalmarks','test_totaltimes',
                                       'test_type','subject_id','test_total_questions').filter(test_type=test_type,subject_id=subject_id)
        print(testList)
        return JsonResponse({'results': list(testList)})
    else:
        return HttpResponse("Problem")

def getTestInfo(request):
    if(request.method == "GET"):
        test_type = request.GET.get('test_type')
        subject_id = request.GET.get('subject_id')
        test_id = request.GET.get('test_id')
        testInfo = Test.objects.values('test_id', 'test_name', 'test_totalmarks', 'test_totaltimes',
                                       'test_type', 'subject_id', 'test_total_questions').filter(test_type=test_type,
                                                                                                 subject_id=subject_id,
                                                                                                 test_id=test_id)
        return JsonResponse({'results': list(testInfo)})

def mcqQuesPaperSubmit(request):
    if (request.method == "POST"):
        test_id=request.POST.get('test_id')
        ret = request.POST.get('obj')
        list = json.loads(ret)
        for value in list:
            McqQuestion(mcq_question_id=value['qid'], mcq_question=value['ques'],
                        mcq_option1=value['option1'], mcq_option2=value['option2'],
                        mcq_option3=value['option3'],mcq_option4=value['option4'],
                        mcq_right_answer=value['rightAnswer'],test_id=Test.objects.get(pk=test_id),approver="Null",
                        datetime=datetime.datetime.now()).save()

        return JsonResponse({'status': '1'})
    else:
        return HttpResponse("Problem")

def isMcqQuesAvailable(request):
    if (request.method == "GET"):
        test_id = request.GET.get('test_id')
        quesList = McqQuestion.objects.values('mcq_question_id', 'mcq_question',
                        'mcq_option1', 'mcq_option2',
                        'mcq_option3','mcq_option4',
                        'mcq_right_answer').filter(test_id=test_id)
        return JsonResponse(list(quesList),safe=False)
    else:
        return HttpResponse("Problem")

def getMcqQuesData(request):
    if (request.method == "GET"):
        test_id = request.GET.get('test_id')
        quesList = McqQuestion.objects.values('mcq_question_id', 'mcq_question').filter(test_id=test_id)
        if not quesList:
            return JsonResponse({'status': 0})
        else:
            return JsonResponse({'status': 1})
    else:
        return HttpResponse("Problem")

def mcqQuesEdit(request):
    test_id=request.GET.get('test_id')
    mcq_ques_id=request.GET.get('mcq_ques_id')
    mcq_ques=request.GET.get('mcq_ques')
    mcq_option1=request.GET.get('mcq_option1')
    mcq_option2=request.GET.get('mcq_option2')
    mcq_option3=request.GET.get('mcq_option3')
    mcq_option4=request.GET.get('mcq_option4')
    mcq_ans=request.GET.get('mcq_answer')
    if(request.method=="GET"):
        McqQuestion.objects.filter(test_id=test_id,
                                   mcq_question_id=mcq_ques_id).update(mcq_question=mcq_ques,
                                                                       mcq_option1=mcq_option1,
                                                                       mcq_option2=mcq_option2,
                                                                       mcq_option3=mcq_option3,
                                                                       mcq_option4=mcq_option4,
                                                                       mcq_right_answer=mcq_ans,
                                                                       approver="Null",
                                                                       datetime=datetime.datetime.now())
        return JsonResponse({'status': 1})
    else:
        return HttpResponse("Problem")
